File: Frontend/src/lib/services/indicators.ts
import { Indicator } from "@/interfaces/entities/indicator";
import { Result } from "@/types/result";
import { invoke } from "@tauri-apps/api/core";

// db
async function insert_indicator(
    env: string,
    indicator: Indicator
): Promise<Result<Indicator>> {
    try {
        indicator.id = 0;

        const data = await invoke<Indicator>("insert_indicator", {
            env,
            indicator,
        });

        return { ok: true, data };
    } catch (error) {
        return { ok: false, error };
    }
}

async function select_indicator(
    env: string,
    indicator?: Indicator
): Promise<Result<Indicator>> {
    try {
        const data = await invoke<Indicator>("select_indicator", {
            env,
            indicator,
        });

        return { ok: true, data };
    } catch (error) {
        return { ok: false, error };
    }
}

async function select_indicators(env: string): Promise<Result<Indicator[]>> {
    try {
        const data = await invoke<Indicator[]>("select_indicators", {
            env,
        });

        return { ok: true, data };
    } catch (error) {
        return { ok: false, error };
    }
}

async function update_indicator(
    env: string,
    indicator: Indicator
): Promise<Result<Indicator>> {
    try {
        const data = await invoke<Indicator>("update_indicator", {
            env,
            indicator,
        });

        return { ok: true, data };
    } catch (error) {
        return { ok: false, error };
    }
}

async function delete_indicator(
    env: string,
    indicator: Indicator
): Promise<Result<number>> {
    try {
        const data = await invoke<number>("delete_indicator", {
            env,
            indicator,
        });

        return { ok: true, data };
    } catch (error) {
        return { ok: false, error };
    }
}

// cache
async function get_active_indicator(
    env: string,
    indicator?: Indicator
): Promise<Result<Indicator>> {
    try {
        const data = await invoke<Indicator>("get_active_indicator", {
            env,
            indicator,
        });

        return { ok: true, data };
    } catch (error) {
        return { ok: false, error };
    }
}

async function get_active_indicators(
    env: string
): Promise<Result<Indicator[]>> {
    try {
        const data = await invoke<Indicator[]>("get_active_indicators", {
            env,
        });

        return { ok: true, data };
    } catch (error) {
        return { ok: false, error };
    }
}

async function get_subscribed_indicators(
    env: string
): Promise<Result<Record<string, number[]> | null>> {
    try {
        const data = await invoke<Record<string, number[]>>(
            "get_subscribed_indicators",
            {
                env,
            }
        );

        return { ok: true, data };
    } catch (error) {
        return { ok: false, error };
    }
}

async function start_active_indicators(env: string): Promise<Result<void>> {
    try {
        await invoke("start_active_indicators", {
            env,
        });

        return { ok: true, data: undefined };
    } catch (error) {
        return { ok: false, error };
    }
}

async function stop_active_indicators(env: string): Promise<Result<void>> {
    try {
        await invoke("stop_active_indicators", {
            env,
        });

        return { ok: true, data: undefined };
    } catch (error) {
        return { ok: false, error };
    }
}

export const indicatorService = {
    // db
    insert: (env: string, indicator: Indicator): Promise<Result<Indicator>> =>
        insert_indicator(env, indicator),

    select: (env: string, indicator?: Indicator): Promise<Result<Indicator>> =>
        select_indicator(env, indicator),

    selectAll: (env: string): Promise<Result<Indicator[]>> =>
        select_indicators(env),

    update: (env: string, indicator: Indicator): Promise<Result<Indicator>> =>
        update_indicator(env, indicator),

    delete: (env: string, indicator: Indicator): Promise<Result<number>> =>
        delete_indicator(env, indicator),

    // cache
    getActive: (
        env: string,
        indicator?: Indicator
    ): Promise<Result<Indicator>> => get_active_indicator(env, indicator),

    getActiveAll: (env: string): Promise<Result<Indicator[]>> =>
        get_active_indicators(env),

    getSubscribed: (
        env: string
    ): Promise<Result<Record<string, number[]> | null>> =>
        get_subscribed_indicators(env),

    startActive: (env: string): Promise<Result<void>> =>
        start_active_indicators(env),

    stopActive: (env: string): Promise<Result<void>> =>
        stop_active_indicators(env),
};

File: Frontend/src-tauri/src/command/indicators.rs
use std::collections::{HashMap, HashSet};

use crate::{
    core::{
        delete_indicator_core, get_active_indicator_core, get_active_indicators_core,
        get_subscribed_indicators_core, insert_indicator_core, select_indicator_core,
        select_indicators_core, start_active_indicators_core, stop_active_indicators_core,
        update_indicator_core,
    },
    models::{
        entities::indicators::Model,
        structs::{CacheIndicators, IndicatorRequest},
    },
};

// db
#[tauri::command]
pub async fn insert_indicator(env: String, indicator: IndicatorRequest) -> Result<Model, String> {
    insert_indicator_core(env, indicator).await
}

#[tauri::command]
pub async fn select_indicator(
    env: String,
    indicator: Option<IndicatorRequest>,
) -> Result<Model, String> {
    select_indicator_core(env, indicator).await
}

#[tauri::command]
pub async fn select_indicators(
    env: String,
    indicator: Option<IndicatorRequest>,
) -> Result<Vec<Model>, String> {
    select_indicators_core(env, indicator).await
}

#[tauri::command]
pub async fn update_indicator(env: String, indicator: IndicatorRequest) -> Result<Model, String> {
    update_indicator_core(env, indicator).await
}

#[tauri::command]
pub async fn delete_indicator(env: String, indicator: IndicatorRequest) -> Result<u64, String> {
    delete_indicator_core(env, indicator).await
}

// cache
#[tauri::command]
pub async fn get_active_indicator(
    env: String,
    indicator: Option<IndicatorRequest>,
) -> Result<Model, String> {
    get_active_indicator_core(env, indicator).await
}

#[tauri::command]
pub async fn get_active_indicators(env: String) -> Result<Option<CacheIndicators>, String> {
    get_active_indicators_core(env).await
}

#[tauri::command]
pub async fn get_subscribed_indicators(
    env: String,
) -> Result<Option<HashMap<String, HashSet<i32>>>, String> {
    get_subscribed_indicators_core(env).await
}

#[tauri::command]
pub async fn start_active_indicators(env: String) -> Result<(), String> {
    start_active_indicators_core(env).await
}

#[tauri::command]
pub async fn stop_active_indicators(env: String) -> Result<(), String> {
    stop_active_indicators_core(env).await
}
